Update Latebound Add Collection Initializer handling

333fred opened the pull request:
Adjusted assert and lateboundaccesskind for latebound
Add method on a collection initializer. Additionally
adjusted compiler-generated detection and added a test.

Fixes https://github.com/dotnet/roslyn/issues/27034.

@dotnet/roslyn-compiler @AlekseyTs @dotnet/analyzer-ioperation for review.

333fred commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.vb:
```diff
@@ -639,6 +640,66 @@ Block[B5] - Exit
+        Implements IEnumerable(Of Integer)
+
+        Sub M(a As Object, b As Object)
+            Dim i = New C From {a}.Add(b)'BIND:"New C From {a}.Add"
```
Ensures that the member reference in the explicit add call is actually explicit in the tree.

etbyrd commented:
@dotnet-bot test windows_debug_vs-integration_prtest please

etbyrd commented:
@dotnet-bot test windows_release_vs-integration_prtest please

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_Latebound.vb:
```diff
@@ -70,7 +70,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-            If receiver IsNot Nothing AndAlso Not receiver.WasCompilerGenerated AndAlso result.Syntax Is receiver.Syntax Then
+            If receiver IsNot Nothing AndAlso
+               ((Not receiver.WasCompilerGenerated AndAlso result.Syntax Is receiver.Syntax) OrElse
+                (receiver.WasCompilerGenerated AndAlso receiver.Syntax.Kind = SyntaxKind.ObjectCreationExpression)) Then
```
>eceiver.Syntax.Kind = SyntaxKind.ObjectCreationExpression [](start = 56, length = 57)

Can we use something other than a syntax check? Are we using special bound nodes as receivers in this case?

333fred replied:
Worked with @AlekseyTs offline to find a better approach.

---
In reply to: [190969490](https://github.com/dotnet/roslyn/pull/27126#discussion_r190969490) [](ancestors = 190969490)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_ObjectInitializer.vb:
```diff
@@ -905,6 +905,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                                                           callerInfoOpt:=topLevelInitializer)
                 invocation.SetWasCompilerGenerated()
 
+                If (invocation.Kind = BoundKind.LateInvocation) Then
```
>( [](start = 19, length = 1)

Parens are not neede in VB. Consider removing, given that the condition is very simple

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_ObjectInitializer.vb:
```diff
@@ -905,6 +905,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 invocation.SetWasCompilerGenerated()
 
+                If (invocation.Kind = BoundKind.LateInvocation) Then
+                    invocation = DirectCast(invocation, BoundLateInvocation).SetLateBoundAccessKind(LateBoundAccessKind.Call)
```
>invocation = DirectCast(invocation, BoundLateInvocation).SetLateBoundAccessKind(LateBoundAccessKind.Call) [](start = 20, length = 105)

Could you please confirm (by examining IL) that we keep calling the same VB runtime helper after this change and also compare against native compiler?

333fred replied:
This change brings us inline with the native compiler, looks like we had a bug here. Previous to this change, we were calling `[Microsoft.VisualBasic]Microsoft.VisualBasic.CompilerServices.NewLateBinding::LateGet`, whereas the native compiler called: `[Microsoft.VisualBasic]Microsoft.VisualBasic.CompilerServices.NewLateBinding::LateCall`. With this change, we also call `LateCall`.

---
In reply to: [190973564](https://github.com/dotnet/roslyn/pull/27126#discussion_r190973564) [](ancestors = 190973564)

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/Binding/BindingCollectionInitializerTests.vb:
```diff
@@ -2007,5 +2007,53 @@ End Class
+        </file>
+    </compilation>
+
+            CompileAndVerify(source, expectedOutput:=<![CDATA[
```
><![CDATA[ [](start = 53, length = 9)

Could probably be just a string literal, since this is just a single line

AlekseyTs commented:
Done with review pass (iteration 1)

etbyrd commented:
@dotnet-bot test windows_debug_vs-integration_prtest please

etbyrd commented:
@dotnet-bot test windows_release_vs-integration_prtest please

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [392144689](https://github.com/dotnet/roslyn/pull/27126#issuecomment-392144689) [](ancestors = 392144689)

333fred commented:
@dotnet/roslyn-compiler need a second review please.

333fred commented:
@dotnet/roslyn-compiler @jaredpar @cston need a second review please

cston commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.vb:
```diff
@@ -639,6 +640,66 @@ Block[B5] - Exit
+        Public Sub Add(l As Long)
+        End Sub
+    End Class
+End Module]]>.Value
```
Minor point: `Mod1` seems unnecessary, here and in other `IOperation` test.

AlekseyTs commented:
@333fred, @cston, @dotnet/roslyn-compiler Please review
